stats: extended maximum number of items in the statistics map

kapi-no opened the pull request:
Drivers, Flash simulator

Extended the maximum possible number of sections that track statistics
for each flash page.

This was previously limited by:
- UTIL_REPEAT macro
- stats.h API

kapi-no committed:
stats: extending api to handle statistics with more than 256 elements

Changed type of the paramater that determines the number of elements in
the stats group. Now it is possible to declared more than 256 elements.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
util: extending UTIL_DEC_N definitions

Extended the UTIL_DEC_N definitions to handle the UTIL_REPEAT macro with
up to 256 repetitions.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

nvlsianpu commented:
cc @ccollins476ad

jukkar commented on include/stats.h:
```diff
@@ -214,8 +214,8 @@ struct stats_hdr {
  */
-void stats_init(struct stats_hdr *shdr, uint8_t size, uint8_t cnt,
-		const struct stats_name_map *map, uint8_t map_cnt);
+void stats_init(struct stats_hdr *shdr, uint8_t size, uint16_t cnt,
```
Shouldn't we have `u16_t` here?

kapi-no replied:
We probably should, but I wanted to keep existing conventions. I think that type alignment for stats.h should be done in a separate PR/commit

nvlsianpu replied:
@jukkar migration form stdint.h is out of scope his PR. Can we kept this as is?

jukkar replied:
Well, it just looks weird that the header file has uint8_t and uint16_t vars, but the .c file uses u8_t and u16_t. In this respect I would say this needs to change.

kapi-no replied:
@jukkar, Stats API now uses Zephyr types

jukkar commented on include/stats.h:
```diff
@@ -255,8 +255,8 @@ int stats_register(const char *name, struct stats_hdr *shdr);
  */
-int stats_init_and_reg(struct stats_hdr *hdr, uint8_t size, uint8_t cnt,
-		       const struct stats_name_map *map, uint8_t map_cnt,
+int stats_init_and_reg(struct stats_hdr *hdr, uint8_t size, uint16_t cnt,
```
and here

kapi-no replied:
fixed

kapi-no committed:
drivers: flash: covering more flash page with stats sections

Extended the maximum possible number of sections that track statistics
for each flash page.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
@nashif or @carlescufi Could you merge?

galak commented on include/misc/util.h:
```diff
@@ -373,6 +373,222 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
 #define UTIL_DEC_38 37
 #define UTIL_DEC_39 38
 #define UTIL_DEC_40 39
+#define UTIL_DEC_41 40
```
Can this be reworked to not require a #define at all?

kapi-no replied:
Could you suggest something specific? How should I go about this?

kapi-no replied:
Could you refactor it in independent PR? It seems you have a clear idea of how to approach this.

kapi-no committed:
stats: changed API to use zephyr types

Changed API of the Statistics module. Now it uses Zephyr types.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no commented:
@galak or @jukkar Could you merge this?

mbolivar reviewed (approved):
Looks sane to me. I think the suggested util.h changes aren't in scope here; it's just extending existing UTIL_REPEAT() functionality as-is, rather than changing how it is implemented in the first place.
